File: src/main/java/com/mredrock/cyxbs/qa/ui/widget/ButtonScrollBehavior.java
package com.mredrock.cyxbs.qa.ui.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.LinearInterpolator;

import androidx.annotation.NonNull;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.view.ViewCompat;

/**
 * Create by yyfbe at 2020-01-26
 * 上滑消失，下滑出现
 */
public class ButtonScrollBehavior extends CoordinatorLayout.Behavior {
    public ButtonScrollBehavior(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public boolean onStartNestedScroll(@NonNull CoordinatorLayout coordinatorLayout, @NonNull View child, @NonNull View directTargetChild, @NonNull View target, int axes, int type) {
        //垂直滚动时监听
        return axes == ViewCompat.SCROLL_AXIS_VERTICAL;
    }

    @Override
    public void onNestedScroll(@NonNull CoordinatorLayout coordinatorLayout, @NonNull View child, @NonNull View target, int dxConsumed, int dyConsumed, int dxUnconsumed, int dyUnconsumed, int type) {
        super.onNestedScroll(coordinatorLayout, child, target, dxConsumed, dyConsumed, dxUnconsumed, dyUnconsumed, type);
        if (dyConsumed > 0) {
            animateOut(child);
        } else if (dyConsumed < 0) {
            animateIn(child);
        }
    }

    //下移出屏幕
    private void animateOut(View view) {
        CoordinatorLayout.LayoutParams layoutParams = (CoordinatorLayout.LayoutParams) view.getLayoutParams();
        int bottomMargin = layoutParams.bottomMargin;
        view.animate().translationY(view.getHeight() + bottomMargin).setInterpolator(new LinearInterpolator()).start();
    }

    //显示到屏幕
    private void animateIn(View view) {
        view.animate().translationY(0).setInterpolator(new LinearInterpolator()).start();
    }
}

File: module_qa/src/main/java/com/mredrock/cyxbs/qa/ui/widget/ButtonScrollBehavior.kt
package com.mredrock.cyxbs.qa.ui.widget

import android.animation.Animator
import android.animation.ValueAnimator
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.View.SCROLL_AXIS_VERTICAL
import androidx.coordinatorlayout.widget.CoordinatorLayout


class ButtonScrollBehavior : CoordinatorLayout.Behavior<View> {

    private var isShow = true
    private var preAnimator: Animator? = null
    private val thresholdValue = 5


    constructor()
    constructor(context: Context?, attrs: AttributeSet?) : super(
            context,
            attrs
    )


    override fun onStartNestedScroll(
            coordinatorLayout: CoordinatorLayout,
            child: View,
            directTargetChild: View,
            target: View,
            axes: Int,
            type: Int
    ): Boolean {
        return SCROLL_AXIS_VERTICAL == axes
    }


    override fun onNestedPreScroll(
            coordinatorLayout: CoordinatorLayout,
            child: View,
            target: View,
            dx: Int,
            dy: Int,
            consumed: IntArray,
            type: Int
    ) {
        if (dy >= thresholdValue && isShow) {
            ValueAnimator.ofFloat(child.translationY, (coordinatorLayout.bottom - child.top).toFloat()).apply {
                addUpdateListener {
                    child.translationY = it.animatedValue as Float
                }
                preAnimator?.cancel()
                preAnimator = this
            }.start()
            isShow = false
        } else if (dy <= -thresholdValue && !isShow) {
            ValueAnimator.ofFloat(child.translationY, 0f).apply {
                addUpdateListener {
                    child.translationY = it.animatedValue as Float
                }
                preAnimator?.cancel()
                preAnimator = this
            }.start()
            isShow = true
        }
    }
}